gpkit/nomials/map.py

```diff
@@ -0,0 +1,206 @@
+"Implements the NomialMap class"
+from collections import defaultdict
+import numpy as np
+from .. import DimensionalityError
+from ..small_classes import HashVector, Quantity, Strings
+from ..small_scripts import mag
+from ..varkey import VarKey
+from .substitution import parse_subs
+
+
+DIMLESS_QUANTITY = Quantity(1, "dimensionless")
+
+
+class NomialMap(HashVector):
+    """Class for efficent algebraic represention of a nomial
+
+    A NomialMap is a mapping between hashvectors representing exponents
+    and their coefficients in a posynomial.
+
+    For example,  {{x : 1}: 2.0, {y: 1}: 3.0} represents 2*x + 3*y, where
+    x and y are VarKey objects.
+    """
+    units = None
+    expmap = None  # used for monomial-mapping postsubstitution; see .mmap()
+    csmap = None   # used for monomial-mapping postsubstitution; see .mmap()
+
+    def units_of_product(self, thing, thing2=None):
+        "Sets units to those of `thing*thing2`"
```

Comment:
not quite what I envisioned, I meant that `units_of_product` should be a helper function that returns units, and `set_units` should still exist and simply set `self.units`. Then when you have a thing and a thing2, you call `self.set_units(_units_of_product(thing, thing2))`.